chore(memory): promote strong-fit contacts drafts for review (#132)
medic/cht-agent: agent-memory/domains/contacts/issues/9090-feat9065-add-cht-datasource-to-support-get-person-by-uuid.md

  - offline-first data access
  - factory pattern (hierarchical datasource)
  - qualifier pattern
  - versioned API namespace (v1)
  - monorepo shared library
related_issues: []
stale: false
---

## Problem

There was no unified, reusable data-access abstraction in cht-core. Code that needed to fetch domain entities (e.g. a person by UUID) did so ad-hoc and differently depending on whether it ran online (server/API) or offline (client), and the existing shared-libs/cht-script-api was narrowly scoped to the rules engine. There was no consistent, type-safe way to expose data access to server code, client code, and custom configuration code (tasks/targets). Beyond a bare by-UUID lookup, there was also no way to fetch a person together with their full parent-contact lineage in one call — callers had to fetch the person and then resolve each ancestor separately (PR #9176).

## Root Cause

shared-libs/cht-script-api was purpose-built for the rules engine and did not provide a general, context-aware (local vs remote) data-access layer, so each consumer reimplemented its own fetching logic and there was no shared contract for retrieving entities like a person.

## Solution

> Converted shared-libs/cht-script-api into a new shared-libs/cht-datasource TypeScript library. Added a DataContext abstraction with getLocalDataContext (offline) and getRemoteDataContext (online), and exposed two usage modes: an imperative hierarchical factory (getDatasource(ctx).v1.person.getByUuid(uuid)) intended to be handed to custom config code, and a declarative curried API (Person.v1.get(ctx)(Qualifier.byUuid(uuid))). Implemented the first entity, Person, with separate local and remote get implementations (Local.Person.v1.get / Remote.Person.v1.get) behind a shared interface — surfaced imperatively as getByUuid — a Qualifier helper (byUuid), and wired the API person controller and the GET /api/v1/person/:uuid route plus the data-context service. Updated consumers (api server-utils/data-context, sentinel purging, admin auth) to the renamed library, and replaced webapp/src/ts/services/cht-script-api.service.ts with cht-datasource.service.ts across its five webapp importers (app.component.ts, auth.service.ts, contact-summary.service.ts, form.service.ts, rules-engine.service.ts) plus their specs; four further webapp/src/ts files in the diff needed only TypeScript casts. A follow-up added a getWithLineage operation to the person module — backed by both local (CouchDB-direct) and remote (HTTP) bindings, with shared lineage-resolution logic factored into local/libs/lineage.ts and contact helpers in libs/contact.ts — and exposed it on the existing GET /api/v1/person/:uuid endpoint (added by this PR) via a new ?with_lineage=true query parameter, returning the person hydrated with its parent hierarchy (PR #9176).

Reviewer: **nitpick (non-blocking):** "four further webapp/src/ts files in the diff needed only TypeScript casts"
holds for three of the four; `edit-report.component.ts` instead extracts a local (`const typeIds = …`)
to dodge a reassignment type error. The five importers and the corrected count are otherwise exact.

```suggestion
Converted shared-libs/cht-script-api into a new shared-libs/cht-datasource TypeScript library. Added a DataContext abstraction with getLocalDataContext (offline) and getRemoteDataContext (online), and exposed two usage modes: an imperative hierarchical factory (getDatasource(ctx).v1.person.getByUuid(uuid)) intended to be handed to custom config code, and a declarative curried API (Person.v1.get(ctx)(Qualifier.byUuid(uuid))). Implemented the first entity, Person, with separate local and remote get implementations (Local.Person.v1.get / Remote.Person.v1.get) behind a shared interface — surfaced imperatively as getByUuid — a Qualifier helper (byUuid), and wired the API person controller and the GET /api/v1/person/:uuid route plus the data-context service. Updated consumers (api server-utils/data-context, sentinel purging, admin auth) to the renamed library, and replaced webapp/src/ts/services/cht-script-api.service.ts with cht-datasource.service.ts across its five webapp importers (app.component.ts, auth.service.ts, contact-summary.service.ts, form.service.ts, rules-engine.service.ts) plus their specs; four further webapp/src/ts files in the diff needed only type-level fixes (three casts, plus one local-variable extraction in edit-report.component.ts). A follow-up added a getWithLineage operation to the person module — backed by both local (CouchDB-direct) and remote (HTTP) bindings, with shared lineage-resolution logic factored into local/libs/lineage.ts and contact helpers in libs/contact.ts — and exposed it on the existing GET /api/v1/person/:uuid endpoint (added by this PR) via a new ?with_lineage=true query parameter, returning the person hydrated with its parent hierarchy (PR #9176).
```

## Code Patterns

> Acquire a DataContext once, then inject it into data functions. Declarative currying: `const getPerson = Person.v1.get(dataContext); await getPerson(Qualifier.byUuid(uuid))` (shared-libs/cht-datasource/src/person.ts). Imperative factory drill-down: `getDatasource(ctx).v1.person.getByUuid(uuid)`. Qualifier builder objects: `byUuid(uuid)` is exported from shared-libs/cht-datasource/src/qualifier.ts, which never names a namespace itself — src/index.ts does the naming with `export * as Qualifier from './qualifier'`, which is why callers — `api/src/controllers/person.js`, at the anchor and on master — write `Qualifier.byUuid(uuid)`. Local/remote strategy split behind a common interface (src/local/person.ts vs src/remote/person.ts). Versioned namespace (lowercase `v1`) for forward-compatible API evolution. Datasource feature template (PR #9176): declare the public operation in src/person.ts, implement parallel local (src/local/person.ts) and remote (src/remote/person.ts) bindings against a shared data-context, and reuse cross-cutting helpers (src/local/libs/lineage.ts, src/libs/contact.ts, src/libs/doc.ts) — `src/libs/contact.ts` is the anchor-era path; #9625 (`5c1f9ba4a`) deleted those 22 lines and folded `NormalizedParent`/`isNormalizedParent`/`Contact` into the new top-level `src/contact.ts`, so follow the template against that file on master; the api/src/controllers/person.js controller delegates to the datasource rather than embedding query logic.

Reviewer: **issue (blocking):** Two of the three names did not move to `src/contact.ts`. At `5c1f9ba4a`,
`NormalizedParent` and `isNormalizedParent` moved to `src/libs/core.ts` (`:153` and `:158` at that
commit, `:247-253` on master). Only `Contact` moved to `src/contact.ts` (`:30` at that commit,
`:40` on master). `src/contact.ts` imports `NormalizedParent` from `./libs/core` (`:3`). A reader
who follows this sentence looks in the wrong file for the type guard. This sentence is new in
`869dd16`.

```suggestion
Acquire a DataContext once, then inject it into data functions. Declarative currying: `const getPerson = Person.v1.get(dataContext); await getPerson(Qualifier.byUuid(uuid))` (shared-libs/cht-datasource/src/person.ts). Imperative factory drill-down: `getDatasource(ctx).v1.person.getByUuid(uuid)`. Qualifier builder objects: `byUuid(uuid)` is exported from shared-libs/cht-datasource/src/qualifier.ts, which never names a namespace itself — src/index.ts does the naming with `export * as Qualifier from './qualifier'`, which is why callers — `api/src/controllers/person.js`, at the anchor and on master — write `Qualifier.byUuid(uuid)`. Local/remote strategy split behind a common interface (src/local/person.ts vs src/remote/person.ts). Versioned namespace (lowercase `v1`) for forward-compatible API evolution. Datasource feature template (PR #9176): declare the public operation in src/person.ts, implement parallel local (src/local/person.ts) and remote (src/remote/person.ts) bindings against a shared data-context, and reuse cross-cutting helpers (src/local/libs/lineage.ts, src/libs/contact.ts, src/libs/doc.ts) — `src/libs/contact.ts` is the anchor-era path; #9625 (`5c1f9ba4a`) deleted those 22 lines; `Contact` moved to the new top-level `src/contact.ts`, and `NormalizedParent`/`isNormalizedParent` moved to `src/libs/core.ts` (`:247-253` on master, imported by `src/contact.ts`), so follow the template against those two files on master; the api/src/controllers/person.js controller delegates to the datasource rather than embedding query logic.
```
